src/main/java/com/ibm/disni/util/MemoryAllocation.java

```diff
@@ -41,47 +38,63 @@ public synchronized static MemoryAllocation getInstance() {
 	}
 	
 	private MemoryAllocation() {
-		table = new ConcurrentHashMap<String, MemBuf>();
+		table = new ConcurrentHashMap<Integer, MemBuf>();
 	}
-	
-	public MemBuf allocate(int size, MemType type, String classname) {
-		if (type == MemType.DIRECT) {
-			String key = generateKey(size, type, classname);
-			MemBuf buf = table.remove(key);
-			if (buf != null) {
-				buf.getBuffer().clear();
-				return buf;
-			}
+
+	private int roundUpSize(int size){
+		// Round up length to the nearest power of two, or the minimum block size
+		if (size < MIN_BLOCK_SIZE) {
+			size = MIN_BLOCK_SIZE;
+		} else {
+			size--;
+			size |= size >> 1;
+			size |= size >> 2;
+			size |= size >> 4;
+			size |= size >> 8;
+			size |= size >> 16;
+			size++;
 		}
-		return _allocate(size, type, classname);
+		return size;
 	}
-	
 
-	public void free(MemBuf memBuf) {
-		if (memBuf.getType() == MemType.DIRECT) {
-			String key = generateKey(memBuf.getBuffer().capacity(),
-					memBuf.getType(), memBuf.getClassname());
-			table.put(key, memBuf);
-		} 
-	}
 
-	private String generateKey(int size, MemType type, String classname) {
-		String key = classname + ":" + size + ":" + type.ordinal();
-		return key;
+	public MemBuf allocate(int size) {
+		size = roundUpSize(size);
+		MemBuf buf = table.remove(size);
+		if (buf == null) {
+			return _allocate(size);
+		} else {
+			return buf;
+		}
 	}
 
-	private MemBuf _allocate(int size, MemType type, String classname) {
-		if (type == MemType.INDIRECT) {
-			ByteBuffer buffer = ByteBuffer.allocate(size);
-			buffer.order(ByteOrder.nativeOrder());
-			long address = 0;
-			return new MemBuf(MemType.INDIRECT, buffer, address, this,
-					classname);
+	/**
+	 * Checks if existing buffer is of the same size as requested - then just clears it.
+	 * Otherwise allocates a new one and puts this buffer to cache.
+	 * @param buf
+	 * @param size
+	 * @return
+	 */
+	public MemBuf cleanOrAllocate(MemBuf buf, int size){
+		if (buf == null) return allocate(size);
+		else if (buf.size() == roundUpSize(size)) {
+			buf.getBuffer().clear();
+			return buf;
 		} else {
-			ByteBuffer buffer = ByteBuffer.allocateDirect(size);
-			buffer.order(ByteOrder.nativeOrder());
-			long address = MemoryUtils.getAddress(buffer);
-			return new MemBuf(MemType.DIRECT, buffer, address, this, classname);
+			free(buf);
+			return allocate(size);
 		}
 	}
+
+	public void free(MemBuf memBuf) {
```

Comment:
I would only offer one way to free a buffer, i.e. not making this method public but using memBuf.free() anywhere. This will help in the future if we allow to have custom MemoryAllocation implementations.

Reply:
But `memBuf.free` need to access this method to put buffer to cache.

Reply:
Yes but it is in the same package, no?

Reply:
Got it. Fixed.

Reply:
Thx